## 1. The NIST Special Publication 800-63B
<p>If you – 50 years ago – needed to come up with a secret password you were probably part of a secret espionage organization or (more likely) you were pretending to be a spy when playing as a kid. Today, many of us are forced to come up with new passwords <em>all the time</em> when signing into sites and apps. As a password <em>inventeur</em> it is your responsibility to come up with good, hard-to-crack passwords. But it is also in the interest of sites and apps to make sure that you use good passwords. The problem is that it's really hard to define what makes a good password. However, <em>the National Institute of Standards and Technology</em> (NIST) knows what the second best thing is: To make sure you're at least not using a <em>bad</em> password. </p>
<p>In this notebook, we will go through the rules in <a href="https://pages.nist.gov/800-63-3/sp800-63b.html">NIST Special Publication 800-63B</a> which details what checks a <em>verifier</em> (what the NIST calls a second party responsible for storing and verifying passwords) should perform to make sure users don't pick bad passwords. We will go through the passwords of users from a fictional company and use python to flag the users with bad passwords. But us being able to do this already means the fictional company is breaking one of the rules of 800-63B:</p>
<blockquote>
  <p>Verifiers SHALL store memorized secrets in a form that is resistant to offline attacks. Memorized secrets SHALL be salted and hashed using a suitable one-way key derivation function.</p>
</blockquote>
<p>That is, never save users' passwords in plaintext, always encrypt the passwords! Keeping this in mind for the next time we're building a password management system, let's load in the data.</p>
<p><em>Warning: The list of passwords and the fictional user database both contain <strong>real</strong> passwords leaked from <strong>real</strong> websites. These passwords have not been filtered in any way and include words that are explicit, derogatory and offensive.</em></p>

In [1]:
# Importing the pandas module
# ... YOUR CODE FOR TASK 1 ...
import pandas as pd

# Loading in datasets/users.csv 
users = pd.read_csv('datasets/users.csv')

# Printing out how many users we've got
# ... YOUR CODE FOR TASK 1 ...
print(users.shape)

# Taking a look at the 12 first users
# ... YOUR CODE FOR TASK 1 ...
users.head(12)

(982, 3)


id         user_name          password
0    1    vance.jennings          joobheco
1    2    consuelo.eaton        0869347314
2    3   mitchel.perkins        fabypotter
3    4    odessa.vaughan         aharney88
4    5    araceli.wilder        acecdn3000
5    6  shawn.harrington           5278049
6    7        evelyn.gay            master
7    8       noreen.hale            murphy
8    9       gladys.ward           lwsves2
9   10   brant.zimmerman  1190KAREN5572497
10  11     leanna.abbott          aivlys24
11  12   milford.hubbard           hubbard

## 2. Passwords should not be too short
<p>If we take a look at the first 12 users above we already see some bad passwords. But let's not get ahead of ourselves and start flagging passwords <em>manually</em>. What is the first thing we should check according to the NIST Special Publication 800-63B?</p>
<blockquote>
  <p>Verifiers SHALL require subscriber-chosen memorized secrets to be at least 8 characters in length.</p>
</blockquote>
<p>Ok, so the passwords of our users shouldn't be too short. Let's start by checking that!</p>

In [5]:
# Calculating the lengths of users' passwords
users['length'] = users['password'].str.len()

# Flagging the users with too short passwords
users['too_short'] = users['length'] < 8

# Counting and printing the number of users with too short passwords
# ... YOUR CODE FOR TASK 2 ...
print(sum(users['too_short']))

# Taking a look at the 12 first rows
# ... YOUR CODE FOR TASK 2 ...
users.head(12)

376


id         user_name          password  length too_short
0    1    vance.jennings          joobheco       8     False
1    2    consuelo.eaton        0869347314      10     False
2    3   mitchel.perkins        fabypotter      10     False
3    4    odessa.vaughan         aharney88       9     False
4    5    araceli.wilder        acecdn3000      10     False
5    6  shawn.harrington           5278049       7      True
6    7        evelyn.gay            master       6      True
7    8       noreen.hale            murphy       6      True
8    9       gladys.ward           lwsves2       7      True
9   10   brant.zimmerman  1190KAREN5572497      16     False
10  11     leanna.abbott          aivlys24       8     False
11  12   milford.hubbard           hubbard       7      True

## 3.  Common passwords people use
<p>Already this simple rule flagged a couple of offenders among the first 12 users. Next up in Special Publication 800-63B is the rule that</p>
<blockquote>
  <p>verifiers SHALL compare the prospective secrets against a list that contains values known to be commonly-used, expected, or compromised.</p>
  <ul>
  <li>Passwords obtained from previous breach corpuses.</li>
  <li>Dictionary words.</li>
  <li>Repetitive or sequential characters (e.g. ‘aaaaaa’, ‘1234abcd’).</li>
  <li>Context-specific words, such as the name of the service, the username, and derivatives thereof.</li>
  </ul>
</blockquote>
<p>We're going to check these in order and start with <em>Passwords obtained from previous breach corpuses</em>, that is, websites where hackers have leaked all the users' passwords. As many websites don't follow the NIST guidelines and encrypt passwords there now exist large lists of the most popular passwords. Let's start by loading in the 10,000 most common passwords which I've taken from <a href="https://github.com/danielmiessler/SecLists/tree/master/Passwords">here</a>.</p>

In [7]:
# Reading in the top 10000 passwords
common_passwords = pd.read_csv('datasets/10_million_password_list_top_10000.txt', header=None, squeeze=True)

# Taking a look at the top 20
#  YOUR CODE FOR TASK 3 
common_passwords.head(20)

0        123456
1      password
2      12345678
3        qwerty
4     123456789
5         12345
6          1234
7        111111
8       1234567
9        dragon
10       123123
11     baseball
12       abc123
13     football
14       monkey
15      letmein
16       696969
17       shadow
18       master
19       666666
Name: 0, dtype: object

## 4.  Passwords should not be common passwords
<p>The list of passwords was ordered, with the most common passwords first, and so we shouldn't be surprised to see passwords like <code>123456</code> and <code>qwerty</code> above. As hackers also have access to this list of common passwords, it's important that none of our users use these passwords!</p>
<p>Let's flag all the passwords in our user database that are among the top 10,000 used passwords.</p>

In [8]:
# Flagging the users with passwords that are common passwords
users['common_password'] = users['password'].isin(common_passwords)

# Counting and printing the number of users using common passwords
# ... YOUR CODE FOR TASK 4 ...
print(sum(users['common_password']))

# Taking a look at the 12 first rows
# ... YOUR CODE FOR TASK 4 ...
users.head(12)

129


id         user_name          password  length too_short common_password
0    1    vance.jennings          joobheco       8     False           False
1    2    consuelo.eaton        0869347314      10     False           False
2    3   mitchel.perkins        fabypotter      10     False           False
3    4    odessa.vaughan         aharney88       9     False           False
4    5    araceli.wilder        acecdn3000      10     False           False
5    6  shawn.harrington           5278049       7      True           False
6    7        evelyn.gay            master       6      True            True
7    8       noreen.hale            murphy       6      True            True
8    9       gladys.ward           lwsves2       7      True           False
9   10   brant.zimmerman  1190KAREN5572497      16     False           False
10  11     leanna.abbott          aivlys24       8     False           False
11  12   milford.hubbard           hubbard       7      True           False

## 5. Passwords should not be common words
<p>Ay ay ay! It turns out many of our users use common passwords, and of the first 12 users there are already two. However, as most common passwords also tend to be short, they were already flagged as being too short. What is the next thing we should check?</p>
<blockquote>
  <p>Verifiers SHALL compare the prospective secrets against a list that contains [...] dictionary words.</p>
</blockquote>
<p>This follows the same logic as before: It is easy for hackers to check users' passwords against common English words and therefore common English words make bad passwords. Let's check our users' passwords against the top 10,000 English words from <a href="https://github.com/first20hours/google-10000-english">Google's Trillion Word Corpus</a>.</p>

In [9]:
# Reading in a list of the 10000 most common words
words = pd.read_csv('datasets/google-10000-english.txt', header=None, squeeze=True)

# Flagging the users with passwords that are common words
users['common_word'] = users['password'].str.lower().isin(words)

# Counting and printing the number of users using common words as passwords
# ... YOUR CODE FOR TASK 5 ...
print(sum(users['common_word']))

# Taking a look at the 12 first rows
# ... YOUR CODE FOR TASK 5 ...
users.head(12)

137


id         user_name          password  length too_short common_password  \
0    1    vance.jennings          joobheco       8     False           False   
1    2    consuelo.eaton        0869347314      10     False           False   
2    3   mitchel.perkins        fabypotter      10     False           False   
3    4    odessa.vaughan         aharney88       9     False           False   
4    5    araceli.wilder        acecdn3000      10     False           False   
5    6  shawn.harrington           5278049       7      True           False   
6    7        evelyn.gay            master       6      True            True   
7    8       noreen.hale            murphy       6      True            True   
8    9       gladys.ward           lwsves2       7      True           False   
9   10   brant.zimmerman  1190KAREN5572497      16     False           False   
10  11     leanna.abbott          aivlys24       8     False           False   
11  12   milford.hubbard           hubbard       7      True           False   

   common_word  
0        False  
1        False  
2        False  
3        False  
4        False  
5        False  
6         True  
7         True  
8        False  
9        False  
10       False  
11       False

## 6. Passwords should not be your name
<p>It turns out many of our passwords were common English words too! Next up on the NIST list:</p>
<blockquote>
  <p>Verifiers SHALL compare the prospective secrets against a list that contains [...] context-specific words, such as the name of the service, the username, and derivatives thereof.</p>
</blockquote>
<p>Ok, so there are many things we could check here. One thing to notice is that our users' usernames consist of their first names and last names separated by a dot. For now, let's just flag passwords that are the same as either a user's first or last name.</p>

In [10]:
# Extracting first and last names into their own columns
users['first_name'] = users['user_name'].str.extract(r'(^\w+)', expand=False)
users['last_name'] = users['user_name'].str.extract(r'(^\w+)', expand=False)

# Flagging the users with passwords that matches their names
users['uses_name'] = (users['first_name'].str.lower() == users['password']) | (users['last_name'].str.lower() == users['password'])

# Counting and printing the number of users using names as passwords
# ... YOUR CODE FOR TASK 6 ...
print(sum(users['uses_name']))

# Taking a look at the 12 first rows
# ... YOUR CODE FOR TASK 6 ...
users.head(12)

22


id         user_name          password  length too_short common_password  \
0    1    vance.jennings          joobheco       8     False           False   
1    2    consuelo.eaton        0869347314      10     False           False   
2    3   mitchel.perkins        fabypotter      10     False           False   
3    4    odessa.vaughan         aharney88       9     False           False   
4    5    araceli.wilder        acecdn3000      10     False           False   
5    6  shawn.harrington           5278049       7      True           False   
6    7        evelyn.gay            master       6      True            True   
7    8       noreen.hale            murphy       6      True            True   
8    9       gladys.ward           lwsves2       7      True           False   
9   10   brant.zimmerman  1190KAREN5572497      16     False           False   
10  11     leanna.abbott          aivlys24       8     False           False   
11  12   milford.hubbard           hubbard       7      True           False   

   common_word first_name last_name uses_name  
0        False      vance     vance     False  
1        False   consuelo  consuelo     False  
2        False    mitchel   mitchel     False  
3        False     odessa    odessa     False  
4        False    araceli   araceli     False  
5        False      shawn     shawn     False  
6         True     evelyn    evelyn     False  
7         True     noreen    noreen     False  
8        False     gladys    gladys     False  
9        False      brant     brant     False  
10       False     leanna    leanna     False  
11       False    milford   milford     False

## 7. Passwords should not be repetitive
<p>Milford Hubbard (user number 12 above), what where you thinking!? Ok, so the last thing we are going to check is a bit tricky:</p>
<blockquote>
  <p>verifiers SHALL compare the prospective secrets [so that they don't contain] repetitive or sequential characters (e.g. ‘aaaaaa’, ‘1234abcd’).</p>
</blockquote>
<p>This is tricky to check because what is <em>repetitive</em> is hard to define. Is <code>11111</code> repetitive? Yes! Is <code>12345</code> repetitive? Well, kind of. Is <code>13579</code> repetitive? Maybe not..? To check for <em>repetitiveness</em> can be arbitrarily complex, but here we're only going to do something simple. We're going to flag all passwords that contain 4 or more repeated characters.</p>

In [11]:
### Flagging the users with passwords with >= 4 repeats
users['too_many_repeats'] = users['password'].str.contains(r'(.)\1')

# Taking a look at the users with too many repeats
# ... YOUR CODE FOR TASK 7 ...
users[users['too_many_repeats']  == True]

id          user_name             password  length too_short  \
0      1     vance.jennings             joobheco       8     False   
2      3    mitchel.perkins           fabypotter      10     False   
3      4     odessa.vaughan            aharney88       9     False   
4      5     araceli.wilder           acecdn3000      10     False   
9     10    brant.zimmerman     1190KAREN5572497      16     False   
11    12    milford.hubbard              hubbard       7      True   
22    23      jenny.woodard              woodard       7      True   
29    30         rene.small             toreze00       8     False   
35    36      dannie.gaines             LCSMCCYF       8     False   
37    38         ervin.dyer  blanchelovesstewart      19     False   
38    39        gus.padilla              wwewwf1       7      True   
41    42         ariel.chen            AASHANNII       9     False   
42    43      jarrett.marsh             96338954       8     False   
43    44     vito.nicholson               225377       6      True   
48    49       junior.burch                diffo       5      True   
49    50     barbra.puckett             madobble       8     False   
56    57       dwayne.hurst              mikeloo       7      True   
61    62      shane.watkins            game67776       9     False   
62    63   dominique.dawson               jfasii       6      True   
63    64     reuben.higgins              sanaang       7      True   
65    66       kieth.barron           NazRam2669      10     False   
69    70      lavern.grimes           stanley#33      10     False   
70    71       bonita.morin        owlsclassrule      13     False   
78    79       roland.mcgee              thx1138       7      True   
88    89         ira.cortez             mmommy22       8     False   
91    92       romeo.gibson               UYHUGG       6      True   
96    97        glen.guerra           0177175959      10     False   
104  105        eve.oconnor      sandra914469025      15     False   
111  112        jacob.boone             CAPRI333       8     False   
112  113      juliette.rios             juliette       8     False   
..   ...                ...                  ...     ...       ...   
863  864     malinda.bright         MUSTANG_2006      12     False   
866  867      lilian.gordon      josue1991101318      15     False   
870  871       berta.holman           5272199335      10     False   
878  879      imogene.oneil           foster1996      10     False   
881  882    fletcher.dillon             81672118       8     False   
882  883       bryan.fisher             marker00       8     False   
887  888   maggie.contreras           0853005163      10     False   
894  895        mike.baxter            pupuubita       9     False   
896  897   suzanne.gallegos             gallegos       8     False   
899  900           ida.wood            gabymoxxa       9     False   
900  901       ashley.house             15goofey       8     False   
914  915      bryant.briggs               briggs       6      True   
916  917     katie.gonzalez           0870575997      10     False   
920  921  alexandria.hinton              hhdmrd.       7      True   
923  924    diann.nicholson             d9990202       8     False   
927  928     roberto.torres              2277272       7      True   
932  933     marlin.fleming              bgeesh1       7      True   
933  934  josephine.morales           mandjuhh89      10     False   
938  939  evangelina.barton             mellamog       8     False   
939  940      lee.hendricks                  lee       3      True   
940  941          zane.bond               pepper       6      True   
941  942     mitch.ferguson               aaaaaa       6      True   
953  954     clarence.britt            28may1997       9     False   
962  963     rico.valentine           ap10172203      10     False   
963  964    angel.jefferson             51183208       8     False   
966  967        

## 8. All together now!
<p>Now we have implemented all the basic tests for bad passwords suggested by NIST Special Publication 800-63B! What's left is just to flag all bad passwords and maybe to send these users an e-mail that strongly suggests they change their password.</p>

In [12]:
# Flagging all passwords that are bad
users['bad_password'] = (users['too_short'])|(users['common_password'])|(users['common_word'])|(users['uses_name'])|(users['too_many_repeats'])

# Counting and printing the number of bad passwords
# ... YOUR CODE FOR TASK 8 ...
print(sum(users['bad_password']))

# Looking at the first 25 bad passwords
# ... YOUR CODE FOR TASK 8 ...
users.head(25)

566


id           user_name          password  length too_short  \
0    1      vance.jennings          joobheco       8     False   
1    2      consuelo.eaton        0869347314      10     False   
2    3     mitchel.perkins        fabypotter      10     False   
3    4      odessa.vaughan         aharney88       9     False   
4    5      araceli.wilder        acecdn3000      10     False   
5    6    shawn.harrington           5278049       7      True   
6    7          evelyn.gay            master       6      True   
7    8         noreen.hale            murphy       6      True   
8    9         gladys.ward           lwsves2       7      True   
9   10     brant.zimmerman  1190KAREN5572497      16     False   
10  11       leanna.abbott          aivlys24       8     False   
11  12     milford.hubbard           hubbard       7      True   
12  13           mamie.fox        mitonguito      10     False   
13  14       jamie.cochran            310356       6      True   
14  15  nathaniel.robinson         angelmajo       9     False   
15  16          lorrie.gay           oZ4k0QE       7      True   
16  17        domingo.dyer           chelsea       7      True   
17  18      martin.pacheco           zvc1939       7      True   
18  19       shelby.massey            nickgd       6      True   
19  20     rosella.barrett        O2gv3LlcfG      10     False   
20  21       karina.morton          dada4943       8     False   
21  22     leticia.sanford          cocacola       8     False   
22  23       jenny.woodard           woodard       7      True   
23  24     brandie.webster          sentry31       8     False   
24  25      sabrina.suarez        OTEL3Q0D8y      10     False   

   common_password common_word first_name  last_name uses_name  \
0            False       False      vance      vance     False   
1            False       False   consuelo   consuelo     False   
2            False       False    mitchel    mitchel     False   
3            False       False     odessa     odessa     False   
4            False       False    araceli    araceli     False   
5            False       False      shawn      shawn     False   
6             True        True     evelyn     evelyn     False   
7             True        True     noreen     noreen     False   
8            False       False     gladys     gladys     False   
9            False       False      brant      brant     False   
10           False       False     leanna     leanna     False   
11           False       False    milford    milford     False   
12           False       False      mamie      mamie     False   
13           False       False      jamie      jamie     False   
14           False       False  nathaniel  nathaniel     False   
15           False       False     lorrie     lorrie     False   
16            True        True    domingo    domingo     False   
17           False       False     martin     martin     False   
18           False       False     shelby     shelby     False   
19           False       False    rosella    rosella     False   
20           False       False     karina     karina     False   
21            True       False    leticia    leticia     False   
22           False       False      jenny      jenny     False   
23           False       False    brandie    brandie     False   
24           False       False    sabrina    sabrina     False   

   too_many_repeats bad_password  
0              True         True  
1             False        False  
2              True         True  
3              True         True  
4              True         True  
5             False         True  
6             False         True  
7             False         True  
8             False         True  
9              True         True  
10            False        False  
11             True         True  
12            False        False  
13            False         True  
14            False        False  
15        

## 9. Otherwise, the password should be up to the user
<p>In this notebook, we've implemented the password checks recommended by the NIST Special Publication 800-63B. It's certainly possible to better implement these checks, for example, by using a longer list of common passwords. Also note that the NIST checks in no way guarantee that a chosen password is good, just that it's not obviously bad.</p>
<p>Apart from the checks we've implemented above the NIST is also clear with what password rules should <em>not</em> be imposed:</p>
<blockquote>
  <p>Verifiers SHOULD NOT impose other composition rules (e.g., requiring mixtures of different character types or prohibiting consecutively repeated characters) for memorized secrets. Verifiers SHOULD NOT require memorized secrets to be changed arbitrarily (e.g., periodically).</p>
</blockquote>
<p>So the next time a website or app tells you to "include both a number, symbol and an upper and lower case character in your password" you should send them a copy of <a href="https://pages.nist.gov/800-63-3/sp800-63b.html">NIST Special Publication 800-63B</a>.</p>

In [0]:
# Enter a password that passes the NIST requirements
# PLEASE DO NOT USE AN EXISTING PASSWORD HERE
new_password = "@<3!4u()"